prov/verbs/src/verbs_domain_xrc.c

```diff
@@ -0,0 +1,576 @@
+		ep->ini_conn->state = FI_IBV_INI_QP_CONNECTING;
+	} else {
+		if (!ep->id->qp) {
+			ep->conn_setup->rsvd_ini_qpn = fi_ibv_reserve_qpn(ep);
+			if (!ep->conn_setup->rsvd_ini_qpn) {
+				VERBS_WARN(FI_LOG_FABRIC,
+					  "rsvd_ini_qpn create err %d\n",
+					  errno);
+				abort();
+			}
+		}
+	}
+
+	assert(ep->ini_conn->ini_qp);
+
+	ep->ibv_qp = ep->ini_conn->ini_qp;
+	dlist_insert_tail(&ep->ini_conn_entry,
+			  &ep->ini_conn->active_list);
+
+	ret = fi_ibv_process_ini_conn(ep, ep->conn_setup->pending_recip,
+				      ep->conn_setup->pending_param,
+				      ep->conn_setup->pending_paramlen);
+	if (ret) {
+		ep->ini_conn->state = last_state;
+		fi_ibv_put_shared_ini_conn(ep);
+
+		/* We need to let the application know that the connect request
+		 * has failed. */
+		fi_ibv_create_shutdown_event(ep);
+	}
+
+	/* Continue to schedule shared connections if the physical connection
+	 * has completed and there are connection requests pending. We could
+	 * implement a throttle here if it is determined that it is better to
+	 * limit the number of outstanding connections. */
+	goto sched_next;
+
+	return;
+}
+
+/* Caller must hold domain:xrc:ini_mgmt_lock */
+int fi_ibv_process_ini_conn(struct fi_ibv_ep *ep,int reciprocal,
+			    void *param, size_t paramlen)
+{
+	struct fi_ibv_xrc_cm_data *cm_data = param;
+	struct rdma_conn_param conn_param = { 0 };
+	int ret;
+
+	assert(ep->ibv_qp);
+
+	if (!reciprocal)
+		fi_ibv_eq_set_xrc_conn_tag(ep);
+
+	fi_ibv_set_xrc_cm_data(cm_data, reciprocal, ep->conn_setup->tag,
+			       ep->eq->pep_port, ep->ini_conn->tgt_qpn);
+	conn_param.private_data = cm_data;
+	conn_param.private_data_len = paramlen;
+	conn_param.responder_resources = RDMA_MAX_RESP_RES;
+	conn_param.initiator_depth = RDMA_MAX_INIT_DEPTH;
+	conn_param.flow_control = 1;
+	conn_param.retry_count = 15;
+	conn_param.rnr_retry_count = 7;
+	conn_param.srq = 1;
+
+	/* Shared connections use reserved temporary QP numbers to
+	 * avoid the appearance of stale/duplicate CM messages */
+	if (!ep->id->qp)
+		conn_param.qp_num = ep->conn_setup->rsvd_ini_qpn->qp_num;
+
+	assert(ep->conn_state == FI_IBV_XRC_UNCONNECTED ||
+	       ep->conn_state == FI_IBV_XRC_ORIG_CONNECTED);
+	ep->conn_state++;
+
+	ret = rdma_connect(ep->id, &conn_param) ? -errno : 0;
+	if (ret) {
+		ret = -errno;
+		VERBS_INFO_ERRNO(FI_LOG_FABRIC, "rdma_connect", errno);
+		ep->conn_state--;
+	}
+	return ret;
+}
+
+int fi_ibv_ep_create_tgt_qp(struct fi_ibv_ep *ep, uint32_t tgt_qpn)
+{
+	struct ibv_qp_open_attr open_attr;
+	struct ibv_qp_init_attr_ex attr_ex;
+	struct fi_ibv_domain *domain = fi_ibv_msg_ep_to_domain(ep);
+	struct ibv_qp *rsvd_qpn;
+
+	assert(ep->tgt_id && !ep->tgt_id->qp);
+
+	/* If a target QP number was specified then open that existing
+	 * QP for sharing. */
+	if (tgt_qpn) {
+		rsvd_qpn = fi_ibv_reserve_qpn(ep);
+		if (!rsvd_qpn) {
+			VERBS_WARN(FI_LOG_FABRIC,
+				   "Create of XRC reserved QPN failed %d\n",
+				   errno);
+			return -errno;
```

Comment:
Use a return code, not errno

Reply:
Modified function being called to return error in status so that it could be returned to the caller of this function.